File: generate.py
"""Small script to generate a static site from parts"""
from os import listdir
from os.path import isfile, join

SITE_PATH = "./site/"
PAGES_PATH = "./pages/"
PAGES = [f for f in listdir(PAGES_PATH) if isfile(join(PAGES_PATH, f))]


def read_part(part_name, pages, default):
    """Read a part of the site"""
    if part_name in pages:
        pages.remove(part_name)
        with open(join(PAGES_PATH, part_name), "r") as header_file:
            return header_file.read()
    else:
        return default


def gen_navbar(pages):
    """Generate the site navbar """
    navbar = "".join(
        [
            '<a target="_self" href="%s">%s</a>\n' % (s, s.rstrip(".html"))
            for s in pages
        ]
    )
    return navbar + "</div></div><div>"


def main():
    """Main function to build the site"""
    footer_str = read_part("footer.html", PAGES, "</div></body></html>")
    header_str = read_part(
        "header.html", PAGES, "<!DOCTYPE html><head></head><body>"
    ) + gen_navbar(PAGES)

    for page in PAGES:
        with open(join(PAGES_PATH, page), "r") as page_file:
            page_str = header_str + page_file.read() + footer_str
            with open(join(SITE_PATH, page), "w") as site_file:
                site_file.write(page_str)

if __name__ == '__main__':
    main()

File: generate.go
package main

import (
    "fmt"
    "io/ioutil"
    "path/filepath"
    "strings"
)

func read_part(part_name string, pages []string, def string) (string, []string, error) {
    for i, page := range pages {
        if page == part_name {
            b, err := ioutil.ReadFile(part_name)
            if err != nil {
                return "", pages, err
            }
            s := string(b[:])
            pages = append(pages[:i], pages[i+1:]...)
            return s, pages, nil
        }
    }
    return def, pages, nil
}

func gen_navbar(pages []string) string {
    navbar := ""
    for _, page := range pages {
        page_path := strings.Replace(page, "pages/", "", -1)
        title := strings.Replace(page_path, ".html", "", -1)
        navbar += fmt.Sprintf("<a target=\"_self\" href=\"%s\">%s</a>\n", page_path, title)
    }
    navbar += "</div></div><div>"
    return navbar
}

func main() {
    site_path := "./site/"
    page_path := "./pages/*"

    pages, err := filepath.Glob(page_path)
    if err != nil {
        return
    }

    footer_str, pages, err := read_part("pages/footer.html", pages, "</div></body></html>")
    if err != nil {
        fmt.Println(err)
    }

    header_str, pages, err := read_part("pages/header.html", pages, "<!DOCTYPE html><head></head><body>")
    header_str += gen_navbar(pages)
    if err != nil {
        fmt.Println(err)
    }

    for _, page := range pages {
        b, err := ioutil.ReadFile(page)
        if err != nil {
            return
        }

        page_str := header_str + string(b[:]) + footer_str
        filename := strings.Replace(page, "pages/", "", -1)

        if err = ioutil.WriteFile(site_path + filename, []byte(page_str), 0666); err != nil {
            fmt.Println(err)
        }
    }
}
